src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs

```diff
@@ -1168,11 +1189,46 @@ private Pipe CreateRequestBodyPipe(uint windowSize)
         }
     }
 
+    public override async ValueTask<IWebTransportSession> AcceptAsync(CancellationToken token)
+    {
+        if (_isWebTransportSessionAccepted)
+        {
+            throw new InvalidOperationException(CoreStrings.AcceptCannotBeCalledMultipleTimes);
+        }
+        _isWebTransportSessionAccepted = true;
+
+        if (!_context.ServiceContext.ServerOptions.EnableWebTransportAndH3Datagrams)
+        {
+            throw new InvalidOperationException(CoreStrings.WebTransportIsDisabled);
+        }
+
+        if (!IsWebTransportRequest)
+        {
+            throw new InvalidOperationException(CoreStrings.FormatFailedToNegotiateCommonWebTransportVersion(WebTransportSession.CurrentSuppportedVersion));
+        }
+
+        // version negotiation
+        var version = WebTransportSession.CurrentSuppportedVersion[WebTransportSession.SecPrefix.Length..];
+
+        // send version negotiation resulting version
+        ResponseHeaders[WebTransportSession.VersionHeaderPrefix] = version;
+        await FlushAsync(token);
+
+        _context.WebTransportSession = _context.Connection!.OpenNewWebTransportSession(this);
```

Comment:
Create the session before responding to the client to avoid races.
```suggestion
        _context.WebTransportSession = _context.Connection!.OpenNewWebTransportSession(this);
        
        // send version negotiation resulting version
        ResponseHeaders[WebTransportSession.VersionHeaderPrefix] = version;
        await FlushAsync(token);
```